mypy/messages.py

```diff
@@ -529,6 +531,20 @@ def has_no_attr(
                     context,
                     code=codes.UNION_ATTR,
                 )
+                if typ_format == '"None"':
+                    if isinstance(context, NameExpr):
+                        var_name = f" {context.name}"
+                    elif isinstance(context, MemberExpr) and isinstance(context.expr, NameExpr):
+                        var_name = f" {context.expr.name}"
+                    elif isinstance(context, MemberExpr) and isinstance(context.expr, MemberExpr):
+                        var_name = f" {get_member_expr_fullname(context.expr)}"
```

Comment:
This branch has some hiccups if the expression isn't strictly a dotted member expression:

```python
class A:
    x: int | None

class B:
    a: A

[A()][0].x.bit_count()  # N: You can use "if None is not None" to guard against a None value
B().a.x.bit_count()     # N: You can use "if None.x is not None" to guard against a None value
```

I think two adjustments would be needed:
1. check for `get_member_expr_fullname` returning `None` (and don't emit the note if it does)
2. fix the case where `get_member_expr_fullname` returns a string with an embedded `"None"`